bridge/src/connection.rs

```diff
@@ -0,0 +1,150 @@
+    InvalidRpc,
+    UnableToConnect(ClientInitError),
+    RequestError(tonic::Status),
+}
+
+impl std::fmt::Display for ConnectionError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            ConnectionError::InvalidUrl(error) => write!(f, "{}", error),
+            ConnectionError::InvalidProtobuf(error) => write!(f, "{}", error),
+            ConnectionError::InvalidConnectionOptions(error) => write!(f, "{}", error),
+            ConnectionError::InvalidRpc => write!(f, "Invalid RPC call"),
+            ConnectionError::UnableToConnect(error) => write!(f, "{}", error),
+            ConnectionError::RequestError(status) => write!(f, "{}", status)
+        }
+    }
+}
+
+impl From<url::ParseError> for ConnectionError {
+    fn from(err: url::ParseError) -> Self {
+        ConnectionError::InvalidUrl(err)
+    }
+}
+
+impl From<prost::DecodeError> for ConnectionError {
+    fn from(err: prost::DecodeError) -> Self {
+        ConnectionError::InvalidProtobuf(err)
+    }
+}
+
+impl From<ClientOptionsBuilderError> for ConnectionError {
+    fn from(err: ClientOptionsBuilderError) -> Self {
+        ConnectionError::InvalidConnectionOptions(err)
+    }
+}
+
+impl From<ClientInitError> for ConnectionError {
+    fn from(err: ClientInitError) -> Self {
+        ConnectionError::UnableToConnect(err)
+    }
+}
+
+impl From<tonic::Status> for ConnectionError {
+    fn from(err: tonic::Status) -> Self {
+        ConnectionError::RequestError(err)
+    }
+}
+
+impl std::error::Error for ConnectionError {}
+
+
+fn rpc_req<P: prost::Message + Default>(bytes: Vec<u8>) -> Result<tonic::Request<P>, prost::DecodeError> {
+    let proto = P::decode(&*bytes)?;
+    Ok(Request::new(proto))
+}
+
+fn rpc_resp<P: prost::Message + Default>(res: Result<tonic::Response<P>, tonic::Status>) -> Result<Vec<u8>, ConnectionError> {
+    Ok(res?.get_ref().encode_to_vec())
+}
+
+macro_rules! rpc_call {
+    ($client:expr, $rpc:ident, $bytes:expr) => {
+        rpc_resp($client.$rpc(rpc_req($bytes)?).await)
+    };
+}
+
+// A Connection is a Client wrapper for making RPC calls
+#[derive(Debug, Clone)]
+pub struct Connection {
+    client: Client
+}
+
+impl Connection {
+    // TODO: Change the interface to accept a full configuration
+    pub async fn connect(host: String) -> Result<Self, ConnectionError> {
+        let url = Url::try_from(&*host)?;
+        let options = ClientOptionsBuilder::default()
+            .identity("testtest".to_string())
+            .worker_binary_id("fakebinaryid".to_string())
+            .target_url(url)
+            .client_name("ruby-sdk".to_string())
+            .client_version("0.0.1".to_string())
+            .build()?;
+
+        let client = options.connect_no_namespace(None, None).await?;
+
+        Ok(Connection { client: client })
+    }
+
+    pub async fn call(&mut self, rpc: &str, bytes: Vec<u8>) -> Result<Vec<u8>, ConnectionError> {
```

Comment:
The user is gonna want to be able to opt-in to retrying calls. And the calls we make from the high-level one we'll always opt in to, but we should allow users making direct calls not to have to.

Reply:
Good point, I'll keep this in mind and add it at a later point 👍